src/multiversioning.zig

```diff
@@ -842,26 +849,31 @@ pub const Multiversion = struct {
         // Zero the atime, so we can compare the rest of the struct directly.
         self.timeout_statx.atime = std.mem.zeroes(os.linux.statx_timestamp);
 
-        if (self.timeout_statx_previous == .err or
+        if (self.timeout_statx_previous == .none or
+            self.timeout_statx_previous == .err or
             (self.timeout_statx_previous == .previous and !stdx.equal_bytes(
             os.linux.Statx,
             &self.timeout_statx_previous.previous,
             &self.timeout_statx,
         ))) {
-            log.info("binary change detected: {s}", .{self.exe_path});
+            if (self.timeout_statx_previous != .none) {
+                log.info("binary change detected: {s}", .{self.exe_path});
+            }
 
             self.stage = .init;
-            self.binary_open();
+            self.binary_open(self.timeout_statx.size);
```

Comment:
check that statx.size is not greater than the size of the buffer we are going to read stuff into